File: clients/typescript/test/contract/aggregate-reservation.test.ts
import { after, before, beforeEach, describe, it } from "node:test";
import assert from "node:assert/strict";

import { InboxWorker } from "../../src/index.ts";
import type { DatabaseHarness } from "./harness.ts";
import { PostgresHarness } from "./postgres-harness.ts";
import { SqlServerHarness } from "./sqlserver-harness.ts";

const SOURCE = "orders";

/**
 * Item 18 of clients/contract-tests.md, from finding F-087. The claim must hold at most one
 * message per aggregate in flight, even under two competing workers.
 */
function aggregateReservation(harness: DatabaseHarness): void {
  describe(`aggregate reservation against ${harness.dialect}`, () => {
    before(async () => {
      await harness.start();
    });

    after(async () => {
      await harness.stop();
    });

    beforeEach(async () => {
      await harness.reset();
    });

    function worker(): InboxWorker {
      return new InboxWorker(harness.connections, {
        source: SOURCE,
        dialect: harness.dialect,
        pollIntervalMs: 50,
        shutdownGraceMs: 10_000,
        batchSize: 4,
      });
    }

    /** Run two workers until `expected` handler calls finish, then stop both. */
    async function runTwoWorkersUntil(expected: number, handler: () => Promise<void>): Promise<void> {
      const stop = new AbortController();
      let handled = 0;
      let settle = (): void => undefined;
      const reached = new Promise<void>((resolve) => {
        settle = resolve;
      });

      const wrapped = async (): Promise<void> => {
        try {
          await handler();
        } finally {
          handled += 1;

          if (handled >= expected) {
            settle();
          }
        }
      };

      const runs = [worker().run(wrapped, stop.signal), worker().run(wrapped, stop.signal)];

      await reached;
      // The failure path runs after the handler returned, so give it room to finish.
      await new Promise((resolve) => setTimeout(resolve, 500));
      stop.abort();
      await Promise.all(runs);
    }

    // Item 18, one aggregate.
    it("never runs two handlers of one aggregate at one time", async () => {
      for (let i = 0; i < 4; i += 1) {
        await harness.insertPending({
          source: SOURCE,
          idempotencyKey: `key-18-single-${i}`,
          payload: "{}",
          aggregateId: "agg-1",
        });
      }

      const windows: Array<[number, number]> = [];

      await runTwoWorkersUntil(4, async () => {
        const start = Date.now();
        await new Promise((resolve) => setTimeout(resolve, 200));
        windows.push([start, Date.now()]);
      });

      assert.equal(windows.length, 4, "the handler must run exactly four times");

      windows.sort((a, b) => a[0] - b[0]);

      for (let i = 1; i < windows.length; i += 1) {
        assert.ok(
          windows[i][0] >= windows[i - 1][1],
          `handler ${i} started at ${windows[i][0]}, before handler ${i - 1} ended at ${windows[i - 1][1]}`,
        );
      }
    });

    // Item 18, two aggregates.
    it("does run two aggregates at one time", async () => {
      for (let i = 0; i < 2; i += 1) {
        await harness.insertPending({
          source: SOURCE,
          idempotencyKey: `key-18-multi-a-${i}`,
          payload: "{}",
          aggregateId: "agg-1",
        });
        await harness.insertPending({
          source: SOURCE,
          idempotencyKey: `key-18-multi-b-${i}`,
          payload: "{}",
          aggregateId: "agg-2",
        });
      }

      let inFlight = 0;
      let peak = 0;

      await runTwoWorkersUntil(4, async () => {
        inFlight += 1;
        peak = Math.max(peak, inFlight);
        await new Promise((resolve) => setTimeout(resolve, 200));
        inFlight -= 1;
      });

      assert.equal(peak, 2, "the peak concurrency must be 2, one per aggregate");
    });
  });
}

aggregateReservation(new PostgresHarness());
aggregateReservation(new SqlServerHarness());

File: clients/go/contract/aggregate_reservation_test.go
package contract_test

import (
	"context"
	"database/sql"
	"fmt"
	"sort"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	queuebox "github.com/alternayte/queuebox/clients/go"
)

// TestOneAggregateNeverRunsTwoHandlers is item 18 of clients/contract-tests.md, from finding
// F-087. The claim must never hand out two rows of one aggregate to run at one time, even under
// two competing workers.
func TestOneAggregateNeverRunsTwoHandlers(t *testing.T) {
	for name, build := range map[string]func() harness{
		"postgresql": func() harness { return &postgresHarness{} },
		"sqlserver":  func() harness { return &sqlServerHarness{} },
	} {
		t.Run(name, func(t *testing.T) {
			h := build()
			h.start(t)
			h.reset(t)

			r := runner{t: t, h: h}
			aggregate := "agg-1"

			for i := range 4 {
				h.insertPending(t, pendingRow{
					source:         source,
					idempotencyKey: fmt.Sprintf("key-18-single-%d", i),
					payload:        "{}",
					aggregateID:    &aggregate,
				})
			}

			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
			defer cancel()

			var (
				mutex   sync.Mutex
				windows [][2]time.Time
			)

			handle := func(_ context.Context, _ queuebox.Message, _ *sql.Tx) error {
				start := time.Now()
				time.Sleep(200 * time.Millisecond)

				mutex.Lock()
				windows = append(windows, [2]time.Time{start, time.Now()})
				mutex.Unlock()

				return nil
			}

			var group sync.WaitGroup

			for range 2 {
				group.Add(1)

				go func() {
					defer group.Done()
					_ = r.worker(t, queuebox.Options{BatchSize: 4}).Run(ctx, handle)
				}()
			}

			deadline := time.Now().Add(90 * time.Second)
			for {
				mutex.Lock()
				done := len(windows)
				mutex.Unlock()

				if done >= 4 || time.Now().After(deadline) {
					break
				}

				time.Sleep(50 * time.Millisecond)
			}

			cancel()
			group.Wait()

			mutex.Lock()
			defer mutex.Unlock()

			if len(windows) != 4 {
				t.Fatalf("the handler ran %d times and it must run 4 times", len(windows))
			}

			sort.Slice(windows, func(i, j int) bool { return windows[i][0].Before(windows[j][0]) })

			for i := 1; i < len(windows); i++ {
				if windows[i][0].Before(windows[i-1][1]) {
					t.Fatalf("handler %d started at %v, before handler %d ended at %v",
						i, windows[i][0], i-1, windows[i-1][1])
				}
			}
		})
	}
}

// TestTwoAggregatesDoRunAtOneTime is the other half of item 18: the reservation must be scoped
// to one aggregate, and must not serialize the whole source.
func TestTwoAggregatesDoRunAtOneTime(t *testing.T) {
	for name, build := range map[string]func() harness{
		"postgresql": func() harness { return &postgresHarness{} },
		"sqlserver":  func() harness { return &sqlServerHarness{} },
	} {
		t.Run(name, func(t *testing.T) {
			h := build()
			h.start(t)
			h.reset(t)

			r := runner{t: t, h: h}

			aggregate1 := "agg-1"
			aggregate2 := "agg-2"

			for i := range 2 {
				h.insertPending(t, pendingRow{
					source:         source,
					idempotencyKey: fmt.Sprintf("key-18-multi-a-%d", i),
					payload:        "{}",
					aggregateID:    &aggregate1,
				})
				h.insertPending(t, pendingRow{
					source:         source,
					idempotencyKey: fmt.Sprintf("key-18-multi-b-%d", i),
					payload:        "{}",
					aggregateID:    &aggregate2,
				})
			}

			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
			defer cancel()

			var (
				inFlight, peak, handled atomic.Int64
			)

			handle := func(_ context.Context, _ queuebox.Message, _ *sql.Tx) error {
				now := inFlight.Add(1)

				for {
					seen := peak.Load()
					if now <= seen || peak.CompareAndSwap(seen, now) {
						break
					}
				}

				time.Sleep(200 * time.Millisecond)
				inFlight.Add(-1)
				handled.Add(1)

				return nil
			}

			var group sync.WaitGroup

			for range 2 {
				group.Add(1)

				go func() {
					defer group.Done()
					_ = r.worker(t, queuebox.Options{BatchSize: 4}).Run(ctx, handle)
				}()
			}

			deadline := time.Now().Add(90 * time.Second)
			for handled.Load() < 4 && time.Now().Before(deadline) {
				time.Sleep(50 * time.Millisecond)
			}

			cancel()
			group.Wait()

			if got := handled.Load(); got != 4 {
				t.Fatalf("the handler ran %d times and it must run 4 times", got)
			}

			if got := peak.Load(); got != 2 {
				t.Fatalf("the peak concurrency was %d and it must be 2, one per aggregate", got)
			}
		})
	}
}
